staging/vhost-device-sound/src/audio_backends/pipewire/test_utils.rs

```diff
@@ -0,0 +1,129 @@
+// Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{
+    path::Path,
+    process::{Child, Command},
+    sync::OnceLock,
+};
+
+use tempfile::{tempdir, TempDir};
+
+/// We want all threads of the same process to share this harness, because cargo
+/// building will block other cargo processes. So build a harness once and share
+/// it with other tests.
+static CELL: OnceLock<PipewireTestHarness> = OnceLock::new();
+
+/// Temporary Dbus session which is killed in drop().
+pub struct DbusSession {
+    pub address: String,
+    pub address_prefix: String,
+}
+
+impl DbusSession {
+    pub fn new(working_dir: &Path) -> Self {
+        let address_prefix = format!("unix:path={}", working_dir.join("dbus").display());
+        let output = Command::new("/usr/bin/dbus-daemon")
+            .args([
+                "--session",
+                "--address",
+                &address_prefix,
+                "--print-address",
+                "--fork",
+            ])
+            .env("DBUS_VERBOSE", "1")
+            .output()
+            .expect("ERROR: dbus-daemon binary not found");
+
+        let mut address = String::from_utf8(output.stdout).unwrap();
+        while address.ends_with('\n') {
+            address.pop();
+        }
+        assert!(
+            address.starts_with(&address_prefix),
+            "Address {} does not start with address prefix `{}`.",
+            address,
+            address_prefix
+        );
+        Self {
+            address,
+            address_prefix,
+        }
+    }
+}
+
+impl Drop for DbusSession {
+    fn drop(&mut self) {
+        println!("INFO: Killing Dbus session {}", self.address);
+        match Command::new("pkill")
+            .arg("-f")
+            .arg(&self.address_prefix)
+            .output()
+        {
+            Err(err) => println!("ERROR: Could not kill dbus-daemon with pkill: {err}"),
+            Ok(output) => {
+                println!(
+                    "TRACE: pkill output: {}",
+                    String::from_utf8_lossy(&output.stdout)
+                );
+                println!(
+                    "TRACE: pkill stderr: {}",
+                    String::from_utf8_lossy(&output.stderr)
+                );
+                println!("TRACE: pkill exit code: {:?}", output.status.code());
+            }
+        }
+    }
+}
+
+/// The pipewire test harness "Singleton". It must only be constructed via
+/// `PipewireTestHarness::new()`.
+pub struct PipewireTestHarness {
```

Comment:
Should we implement `drop()` also here and kill `pipewire`?

Reply:
From https://doc.rust-lang.org/std/process/struct.Child.html#

"There is no implementation of [Drop](https://doc.rust-lang.org/std/ops/trait.Drop.html) for child processes, so if you do not ensure the Child has exited then it will continue to run, even after the Child handle to the child process has gone out of scope."

So I think we should add `drop()` and call `self.pipewire_child.kill()`